src/Operation/InsertOne.php

```diff
@@ -103,8 +105,52 @@ public function execute(Server $server)
         $insertedId = $bulk->insert($this->document);
 
         $writeConcern = isset($this->options['writeConcern']) ? $this->options['writeConcern'] : null;
-        $writeResult = $server->executeBulkWrite($this->databaseName . '.' . $this->collectionName, $bulk, $writeConcern);
+
+        try {
+            $writeResult = $server->executeBulkWrite($this->databaseName . '.' . $this->collectionName, $bulk, $writeConcern);
+
+            return new InsertOneResult($writeResult, $insertedId);
+        } catch (BulkWriteException $e) {
+            if ($this->isBulkWriteExceptionDuplicateKeyError($e)) {
+                throw DuplicateKeyException::fromBulkWriteException($e);
+            }
+
+            throw $e;
+        }
 
         return new InsertOneResult($writeResult, $insertedId);
     }
+
+    /**
+     * Determine if a BulkWriteException is a DuplicateKeyException.
+     *
+     * We only wrap a BulkWriteException if it identifies as a duplicate key
+     * error or contains exactly one write error that identifies as a duplicate
+     * key error. It must not contain a write concern error.
```

Comment:
I think this makes sense, but it's not based on prior art that I know of.

Reply:
/cc @alcaeus @ajdavis

Reply:
Interesting to [compare with mongo-php-adapter](https://github.com/alcaeus/mongo-php-adapter/blob/1.1.3/lib/Alcaeus/MongoDbAdapter/ExceptionConverter.php#L44), which simply checks the first write error's code (if any), without caring about additional write errors or a possible write concern error.

Reply:
IIRC, the logic to only check the first write result was written to cover the case where a single insert would cause a duplicate key exception. I'd have to take a look at the legacy driver to see its behavior during bulk writes to see its behavior when multiple (or any other than the first) inserts result in a duplicate key error.

Reply:
> it may indeed be safer to make sure we only get one write result so we don't hide errors from other bulk write operations.

Indeed, my intention with this logic was that we should only report a DuplicateKeyException if we can assert that the BulkWriteException contains nothing more than a single duplicate key error. Anything else (e.g. multiple write errors, a write concern error) is ambiguous and is best left to the user to inspect on their own.

Reply:
> I'd have to take a look at the legacy driver to see its behavior during bulk writes to see its behavior when multiple (or any other than the first) inserts result in a duplicate key error.

It would appear that `MongoWriteBatch::execute()` reports all write and write concern errors as a MongoWriteConcernException (see: [here](https://github.com/mongodb/mongo-php-driver-legacy/blob/1.6.16/batch/write.c#L427)). This jives with the [PHP.net documentation](http://php.net/manual/en/mongowritebatch.execute.php).

MongoDuplicateKeyException is only used in [`php_mongo_cursor_throw()`](https://github.com/mongodb/mongo-php-driver-legacy/blob/1.6.16/cursor_shared.c#L540), which is called from many places where a command response is read from a cursor. The most relevant places might be [`php_mongo_trigger_error_on_command_failure()`](https://github.com/mongodb/mongo-php-driver-legacy/blob/1.6.16/php_mongo.c#L649) and [`php_mongo_trigger_error_on_gle()`](https://github.com/mongodb/mongo-php-driver-legacy/blob/1.6.16/php_mongo.c#L695), but that won't come up for write commands, as any duplicate key error code would only show up within the `writeErrors` array of the command response and GLE isn't used anymore.